lightning/src/ln/chan_utils.rs

```diff
@@ -691,58 +776,425 @@ impl HolderCommitmentTransaction {
+	to_broadcaster_value_sat,
+	to_countersignatory_value_sat,
+	feerate_per_kw,
+	htlcs,
+	keys,
+	built
+});
+
+impl CommitmentTransaction {
+	/// Construct an object of the class while assigning transaction output indices to HTLCs.
 	///
-	/// The returned Vec has one entry for each HTLC, and in the same order. For HTLCs which were
-	/// considered dust and not included, a None entry exists, for all others a signature is
-	/// included.
-	pub fn get_htlc_sigs<T: secp256k1::Signing + secp256k1::Verification>(&self, htlc_base_key: &SecretKey, counterparty_selected_contest_delay: u16, secp_ctx: &Secp256k1<T>) -> Result<Vec<Option<Signature>>, ()> {
-		let txid = self.txid();
-		let mut ret = Vec::with_capacity(self.per_htlc.len());
-		let holder_htlc_key = derive_private_key(secp_ctx, &self.keys.per_commitment_point, htlc_base_key).map_err(|_| ())?;
-
-		for this_htlc in self.per_htlc.iter() {
-			if this_htlc.0.transaction_output_index.is_some() {
-				let htlc_tx = build_htlc_transaction(&txid, self.feerate_per_kw, counterparty_selected_contest_delay, &this_htlc.0, &self.keys.broadcaster_delayed_payment_key, &self.keys.revocation_key);
-
-				let htlc_redeemscript = get_htlc_redeemscript_with_explicit_keys(&this_htlc.0, &self.keys.broadcaster_htlc_key, &self.keys.countersignatory_htlc_key, &self.keys.revocation_key);
-
-				let sighash = hash_to_message!(&bip143::SigHashCache::new(&htlc_tx).signature_hash(0, &htlc_redeemscript, this_htlc.0.amount_msat / 1000, SigHashType::All)[..]);
-				ret.push(Some(secp_ctx.sign(&sighash, &holder_htlc_key)));
+	/// Populates HTLCOutputInCommitment.transaction_output_index in htlcs_with_aux.
+	///
+	/// The generic T allows the caller to match the HTLC output index with auxiliary data.
+	/// This auxiliary data is not stored in this object.
+	///
+	/// Only include HTLCs that are above the dust limit for the channel.
+	pub fn new_with_auxiliary_htlc_data<T>(commitment_number: u64, to_broadcaster_value_sat: u64, to_countersignatory_value_sat: u64, keys: TxCreationKeys, feerate_per_kw: u32, htlcs_with_aux: &mut Vec<(HTLCOutputInCommitment, T)>, channel_parameters: &DirectedChannelTransactionParameters) -> CommitmentTransaction {
+		// Sort outputs and populate output indices while keeping track of the auxiliary data
+		let (outputs, htlcs, _) = Self::internal_build_outputs(&keys, to_broadcaster_value_sat, to_countersignatory_value_sat, htlcs_with_aux, channel_parameters).unwrap();
+
+		let (obscured_commitment_transaction_number, txins) = Self::internal_build_inputs(commitment_number, channel_parameters);
+		let transaction = Self::make_transaction(obscured_commitment_transaction_number, txins, outputs);
+		let txid = transaction.txid();
+		CommitmentTransaction {
+			commitment_number,
+			to_broadcaster_value_sat,
+			to_countersignatory_value_sat,
+			feerate_per_kw,
+			htlcs,
+			keys,
+			built: BuiltCommitmentTransaction {
+				transaction,
+				txid
+			},
+		}
+	}
+
+	#[cfg(feature = "_test_utils")]
```

Comment:
The scripts seem mostly useful for tests, not so much in production.  This gate here (and below) saves an allocation in the production path.  Could still ungate it if you think it's not a big deal, but I thought you wanted it closer to optimal for memory allocations...

Reply:
What is the testing use-case? I presume just iterating over the HTLCs and building+hashing the redeemscripts is definitely useful in production, but presumably its rather trivial to do so with `htlcs()` instead of trying to rebuild them from the scripts?

Definitely prefer to be careful with in-loop memory allocations (just my bias because Bitcoin Core tends to use almost 2x the expected memory on heap fragmentation alone), but the scripts stuff is only one allocation (and a ton of memcpys), so its probably not the end of the world.

Reply:
I'm thinking either a sweep tx fails to confirm because of wrong redeem script in the witness or a custom signer uses the wrong redeem script and the signature doesn't verify.  These are issues I personally ran into.  Those cases are a bit hard to investigate with just the script hash to go by.

That said, the low-level script utility functions are pub, so there is that path, although less convenient.

Reply:
Right, I guess the other concern is `_test_utils` is intended really for our own internal testing - specifically testing of things in the other crates which want to use some of the `lightning` test bits. I don't really think we should be exposing anything which we intend any downstream users to use via it.

Reply:
Well, lightning-signer actually does use `TestFeeEstimator` and neighbors for its own integration test, so we are (mis)using this internal flag.

I am introducing a public `test_utils` feature flag in a new commit I just pushed, and maybe we can look into moving some of these to this flag.

Alternatively, I am also OK with removing this for now and reintroducing something later that doesn't clutter the production path.

Reply:
Right, I figure just drop it for now - if we want to have external features that expose additional things we probably need to figure out the documentation story around them (ie getting docs.rs to build with them and documenting them appropriately) as well.

Reply:
Taken out.